File: 1942-the-number-of-the-smallest-unoccupied-chair/1942-the-number-of-the-smallest-unoccupied-chair.java
class Solution {

    public int smallestChair(int[][] times, int targetFriend) {
        List<int[]> events = new ArrayList<>(); // To store both arrival and leave events

        // Populate events with {arrival, friendIndex} and {leave, -friendIndex}
        for (int i = 0; i < times.length; i++) {
            events.add(new int[] { times[i][0], i }); // Friend arrives
            events.add(new int[] { times[i][1], ~i }); // Friend leaves
        }

        Collections.sort(events, (a, b) -> a[0] - b[0]); // Sort events by time

        PriorityQueue<Integer> availableChairs = new PriorityQueue<>(); // Min-heap for available chairs
        PriorityQueue<int[]> occupiedChairs = new PriorityQueue<>(
            (a, b) -> a[0] - b[0]
        ); // Min-heap to track when chairs will be vacated

        for (int i = 0; i < times.length; i++) {
            availableChairs.add(i); // Initially all chairs are free
        }

        for (int[] event : events) {
            int time = event[0];
            int friendIndex = event[1];

            // Free up chairs when friends leave
            while (
                !occupiedChairs.isEmpty() && occupiedChairs.peek()[0] <= time
            ) {
                availableChairs.add(occupiedChairs.poll()[1]);
            }
            if (friendIndex >= 0) { // Friend arrives
                int chair = availableChairs.poll();
                if (friendIndex == targetFriend) {
                    return chair;
                }
                occupiedChairs.add(new int[] { times[friendIndex][1], chair }); // Mark when the chair will be vacated
            }
        }

        return -1; // Should never reach here
    }
}

File: 1942-the-number-of-the-smallest-unoccupied-chair/1942-the-number-of-the-smallest-unoccupied-chair.js
/**
 * @param {number[][]} times
 * @param {number} targetFriend
 * @return {number}
 */
var smallestChair = function(times, targetFriend) {
    const totalPersons = times.length;
    let availableChairs = new PriorityQueue({ compare : (a, b) => {
       return a - b;
    }});
    let allocatedChairs = new PriorityQueue({compare : (a, b) => {
        if(a[1] == b[1]) {
            return a[2] - b[2];
        }
        return a[1] - b[1];
     }});
    //init
    for(let i = 0; i < totalPersons; i++) {
        availableChairs.enqueue(i);
    }
    times.map((value,index) => {
        return value.push(index);
    })
    
    times.sort((a,b) => {
        let inTime1 = a[0], inTime2 = b[0];
        if(inTime1 == inTime2) {
            let outTime1 = a[0], outTime2 = b[0];
            return outTime1 - outTime2;
        }
        return inTime1 - inTime2;
    });
    //console.log(times);
    for(let i = 0; i < totalPersons; i++) {
        let inTime = times[i][0], outTime = times[i][1];
        while(allocatedChairs.size() && allocatedChairs.front()[1] <= inTime) {
            let chair = allocatedChairs.front()[3];
            allocatedChairs.dequeue();
            availableChairs.enqueue(chair);
        }
        let chair = availableChairs.dequeue();
        times[i].push(chair);
        allocatedChairs.enqueue(times[i]);
    }
    for(let i = 0; i < totalPersons; i++) {
        if(times[i][2] == targetFriend) {
            return times[i][3];
        }
    }

    //console.log(times);
    return 0;
};